Support group concat function

Jasmine-ge opened the pull request:
Calculates a concatenated string from a group of strings, optionally separated by a delimiter, and optionally limited by a maximum number of elements.

**Syntax**
group_concat (expression [, delimiter [, limit ] ])

**Arguments**

- expression — The expression or column name that outputs strings to be concatenated.
- delimiter — A string that will be used to separate concatenated values. This parameter is optional and defaults to an empty string or delimiter from parameters if not specified.

**Parameters**

- delimiter — A string that will be used to separate concatenated values. This parameter is optional and defaults to an empty string if not specified.
- limit — A positive integer specifying the maximum number of elements to concatenate. If more elements are present, excess elements are ignored. This parameter is optional.

**Returned value**

- Returns a string consisting of the concatenated values of the column or expression. If the group has no elements or only null elements, and the function does not specify a handling for only null values, the result is a nullable string with a null value.

Please write user-readable short description of the changes:
Port CH PRs and update the obvious refactor (utils, format, etc):
https://github.com/ClickHouse/ClickHouse/pull/65009
https://github.com/ClickHouse/ClickHouse/pull/69455
https://github.com/ClickHouse/ClickHouse/pull/72540

close #889

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/timeplus-io/proton?pullRequest=898) <br/>All committers have signed the CLA.

Jasmine-ge committed:
init

Rearranged includes

groupConcat consistency

Change the logic

Change the logic to use analyzer

Update AggregateFunctionGroupConcat.h

Fix run issues

Remove templates from GroupConcatData

support group concat function

yl-lisen commented on src/Interpreters/ExpressionAnalyzer.cpp:
```diff
@@ -277,6 +277,30 @@ void tryTranslateToParametricAggregateFunction(
     }
+    else if (lower_name == "group_concat")
+    {
+        if (arguments.size() > 3 || arguments.size() < 1)
+        {
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Incorrect number of parameters for aggregate function {}, should be 0, 1 or 2, got: {}", node->name, parameters.size());
+        }
+        else if (arguments.size() == 3)
```
```suggestion
        if (arguments.size() > 3 || arguments.size() < 1)
            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
                "Incorrect number of parameters for aggregate function {}, should be 0, 1 or 2, got: {}", node->name, parameters.size());

        /// ...
```

Jasmine-ge committed:
fix coding problems and add notes
